Optimisation and Better functions

UltraStudioLTD opened the pull request:
Fixed few performance and functional issues

UltraStudioLTD commented:
I currently only have Termux (Linux terminal environment for Android) so I can't test Windows functions therefore they are experimental

UltraStudioLTD commented:
Also for some reason click and release functions are not registered on my Termux during testing, although scrolling does. I don't know what is reason so I will check code again and test each parts. If it is working on other OSes than it is okay 👍

UltraStudioLTD commented:
Checked original code and it registers click/releases so it means something is wrong in my code. But I don't see any problems.

bczsalba commented:
Hey! 

May I ask what's up with all the formatting changes you've also included? I appreciate the sentiment of the PR, but there is hundreds of odd changes to files you didn't touch otherwise.

There is no contribution guide as of yet, but if there was it would mention the use of both `pylint` and `black`. I haven't actually run your code yet as I haven't had time, but running it through `pylint` gave a good couple of results.

UltraStudioLTD commented:
I used ```black``` (formatting) and ```isort``` (sorting imports). I didn't used `pylint` but used `mypy` which yielded errors but code itself was working. I only manually refactored `ansi_interface.py`. I left `translate_mouse` and `report_mouse` functions currently as it is but will refactor them as soon as I will have time. I also modified few functions that used `tput` to use ANSI sequences instead so they would be cross-platform. Also I changed `set_echo` and `unset_echo` so they would be cross-platform too (including Windows) but since I tested it on Linux it is currently experimental and only in theory. I also added `TerminalModes` class which has raw mode function which you can use like this:

```python
with terminal_modes.NoEcho():
    # Starts `no_echo` mode
    # Do something
# Exits `no_echo` mode
```

I also refactored `Color` class and hex translation is now better

UltraStudioLTD commented:
Also as I see your code is not optimized and thus decreasing performance. Please don't use `_` in places where you aren't intending function, variable or class to be private.

UltraStudioLTD commented:
Please follow **PEP 8** rules

bczsalba commented:
In what places do I not follow the `_` internal notation? Pylint yells at you if you access any underscore name from outside it's intended scope, and I make very certain it doesn't happen. And even if I _did_ misuse this notation, that does **not** mean the code is unoptimized.

Also, black and pylint are made to follow PEP8, and my code fully compiles with it.

bczsalba commented on pytermgui/context_managers.py:
```diff
@@ -8,26 +8,15 @@
+from os import name
+from typing import Any, Callable, Generator, List, Union
+
+from .ansi_interface import (MouseEvent, cursor_up, hide_cursor,
+                             is_interactive, print_to, report_mouse,
+                             restore_cursor, save_cursor, set_alt_buffer,
+                             set_echo, show_cursor, translate_mouse,
+                             unset_alt_buffer, unset_echo)
```
This is the type of formatting change I don't quite understand. This is a file you didn't actually touch, and for me Black changes this into a much more readable format.

bczsalba commented:
So I looked at the code some more, and I do enjoy a lot of the changes. My problem is that this PR is massive, and there is quite a few things I do not enjoy about it, and without having them in granular commits it is very hard to pick apart what to add and what not to. 

I particularly enjoy the context manager functionalities added to the terminal, but I feel like it can be merged into the `_Terminal` class, as it doesn't really do all the much at the moment. I also like the idea for the `_WinSIGWINCH` implementation, though I'm not quite sure on the use of async for it (as all other parts of the code are threaded instead).

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -664,39 +750,39 @@ def report_mouse(
+        sys.stdout.write("\x1b[?1001")
 
     elif event == "press_hold":
-        _stdout.write("\x1b[?1002")
+        sys.stdout.write("\x1b[?1002")
 
     elif event == "hover":
-        _stdout.write("\x1b[?1003")
+        sys.stdout.write("\x1b[?1003")
 
     else:
         raise NotImplementedError(f"Mouse report event {event} is not supported!")
 
-    _stdout.write("l" if stop else "h")
+    sys.stdout.write("l" if stop else "h")
 
     if method == "decimal_utf8":
-        _stdout.write("\x1b[?1005")
+        sys.stdout.write("\x1b[?1005")
 
     elif method == "decimal_xterm":
-        _stdout.write("\x1b[?1006")
+        sys.stdout.write("\x1b[?1006")
 
     elif method == "decimal_urxvt":
-        _stdout.write("\x1b[?1015")
+        sys.stdout.write("\x1b[?1015")
```
I'm neutral to these changes. The reason `sys.stdout` is imported as `_stdout` is a remnant from before I had `__all__` set up, as I didn't want a star import to also pull in `sys`. In this current form this works fine as well.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -723,114 +809,132 @@ def translate_mouse(code: str, method: str) -> list[MouseEvent | None] | None:
     events: list[MouseEvent | None] = []
     for sequence in code.split("\x1b"):
         if len(sequence) == 0:
             continue
-
         matches = list(pattern.finditer(sequence))
         if matches == []:
             return None
-
         for match in matches:
             identifier, *pos, release_code = match.groups()
 
             # decimal_xterm uses the last character's
             # capitalization to signify press/release state
             if len(release_code) > 0 and identifier == "0":
                 identifier += release_code
-
             if identifier in mapping:
                 action = mapping[identifier]
                 assert isinstance(action, MouseAction)
-
                 events.append(MouseEvent(action, (int(pos[0]), int(pos[1]))))
                 continue
-
             events.append(None)
```
These linebreaks are there for readability. I do not agree with removing them.

UltraStudioLTD replied:
Ok

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -723,114 +809,132 @@ def translate_mouse(code: str, method: str) -> list[MouseEvent | None] | None:
-    method."""
+def print_to(
+    pos: tuple[int, int],
+    *objects: Any,
+    sep: str = " ",
+    end: str = " ",
+    file: Optional[IO[str]] = None,
+    flush: bool = True,
+) -> None:
+    """
+    Print text to given `pos`.
 
+    This passes through all arguments (except for `pos`) to the `print` method.
+    """
     move_cursor(pos)
-    print(*args, **kwargs, end="", flush=True)
+    print(*objects, sep=sep, end=end, file=file, flush=flush)
```
No need to individually take all `kwargs` of `print` as just unpacking `**kwargs` into it does the same thing. People using this relatively low level (in terms of this library) function will more than likely know the signature of `print` anyways, so this is just redundant.

UltraStudioLTD replied:
I just followed `print`'s signature

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -723,114 +809,132 @@ def translate_mouse(code: str, method: str) -> list[MouseEvent | None] | None:
+    """
+    Return text in bold
+    """
+    ending = "\x1b[0m" if reset_style else ""
+    return f"\x1b[1m{text}{ending}"
```
I'm fine with these changes, but not happy with the docstring & linebreak situation. Also, if we have `DECORATION_MODES` as a global, these functions could all be simplified to a table lookup.

UltraStudioLTD replied:
It will be better to use less calls since it consumes memory and CPU time so I think inline variable is better

bczsalba commented on examples/krm:
```diff
@@ -12,12 +12,11 @@ The currently supported commands are:
 - kitty @ resize-os-window --height {height}: Set OS window's height
 """
 
+from argparse import ArgumentParser, Namespace
```
The project organizes its imports by length, not alphabetically.

UltraStudioLTD replied:
I didn't organized imports manually. I used `isort` as I said before

bczsalba commented on pytermgui/__init__.py:
```diff
@@ -13,22 +13,22 @@
 
+from .ansi_interface import *
+from .context_managers import alt_buffer, cursor_at, mouse_handler
 from .enums import *
-from .parser import *
-from .widgets import *
+from .exceptions import *
+from .file_loaders import *
 from .helpers import *
+from .input import getch, keys
 from .inspector import *
+from .parser import *
 from .serializer import *
-from .exceptions import *
-from .file_loaders import *
-from .ansi_interface import *
+from .widgets import *
```
Same as above.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -2,27 +2,36 @@
 
 from __future__ import annotations
 
+import asyncio
```
Not really sure on the usage of `async` functionality for this one use, as all other parts of the library use threads instead.

UltraStudioLTD replied:
It is for `WinSIGWINCH`

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -69,27 +78,53 @@
     "overline",
 ]
 
+_HEX_REGEX: re.Pattern = re.compile(r"#?[0-9a-zA-Z]{3,6}", re.IGNORECASE)
+"""
+Regex to validate HEX
+"""
+
+_CURSOR_POS_REGEX: re.Pattern = re.compile(r".*\x1b\[(?P<y>\d*);(?P<x>\d*)R")
+"""
+Regex to parse cursor query string
+"""
+
+DECORATION_MODES: dict[str, int] = {
+    "reset": 0,
+    "bold": 1,
+    "dim": 2,
+    "italic": 3,
+    "underline": 4,
+    "blink": 5,
+    "inverse": 7,
+    "invisible": 8,
+    "strikethrough": 9,
+    "overline": 53,
+}
+"""
+Dictionary of decoration modes
+"""
```
As I explain above, I appreciate these changes, as well as the documentation added to them. Not happy with the docstring style though, see below.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -69,27 +78,53 @@
+_HEX_REGEX: re.Pattern = re.compile(r"#?[0-9a-zA-Z]{3,6}", re.IGNORECASE)
+"""
+Regex to validate HEX
+"""
+
+_CURSOR_POS_REGEX: re.Pattern = re.compile(r".*\x1b\[(?P<y>\d*);(?P<x>\d*)R")
+"""
+Regex to parse cursor query string
+"""
+
+DECORATION_MODES: dict[str, int] = {
+    "reset": 0,
+    "bold": 1,
+    "dim": 2,
+    "italic": 3,
+    "underline": 4,
+    "blink": 5,
+    "inverse": 7,
+    "invisible": 8,
+    "strikethrough": 9,
+    "overline": 53,
+}
+"""
+Dictionary of decoration modes
+"""
+
 
 class Color:
-    """Class to store various color utilities
+    """
+    Class to store various color utilities
 
-    Two instances of this class are provided, `foreground`
-    and `background`. The difference between these is the color
-    layer they operate on.
+    Two instances of this class are provided, `foreground` and `background`. The difference between these is the color layer they operate on.
 
-    To use this class you should call either instances with some
-    data type representing a color. The following patterns are supported:
+    To use this class you should call either instances with some data type representing a color. The following patterns are supported:
```
Why make these lines so long?

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -69,27 +78,53 @@
     - `int`: 0-256 terminal colors
     - `str`: Name of one of the registered named colors. See `Color.names`.
-    - `#rrggbb`: RGB hex string. Note: alpha values are not supported.
+    - `#rrggbb` or '#rgb': RGB hex string. Note: alpha values are not supported.
```
I appreciated the addition of shorthand hex colors.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
-        `layer` can be either 0 or 1. This value determines whether the instance
-        will represent foreground or background colors."""
+        `layer` can be either 0 or 1. This value determines whether the instance will represent foreground or background colors.
+        """
```
Unnecessary changes to the doctring format and length. Why?

UltraStudioLTD commented:
`signal`s are asynchronous not threaded so I decided to use `async`hronous implementation. About `TerminalModes` class being merged with `Terminal` (not `_Terminal`. Please remove `_` during imports and in variables/classes which are intended to be public) class is possible but I prefer them being separated since I am also planning to add few other modes context managers and it would be more readable for them to be in separate class. Also I tested code and it's working. (Don't know `hover` and `drag` events since I don't have physical mouse, only touchscreen)

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
 
-        if layer not in [0, 1]:
-            raise NotImplementedError(f"Layer {layer} can only be one of [0, 1].")
+        if layer not in (0, 1):
+            raise NotImplementedError(f"Layer {layer} can only be one of (0, 1).")
```
This makes no functional difference as far as I can tell. Using a set might have made a tiny bit more sense semantically, but the difference between lists and tuples aren't important in this case.

UltraStudioLTD replied:
Tuples are more memory efficient. It is explained in few StackOverflow Q/A

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
     @staticmethod
     def translate_hex(color: str) -> tuple[int, int, int]:
-        """Translate hex string of format #RRGGBB into an RGB tuple of integers"""
+        """
+        Translate hex string of format '#RGB' or '#RRGGBB' into an 'RGB' tuple of integers
+        """
```
Same issue with docstrings.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
-        for i in (0, 2, 4):
-            color_hex = color[i : i + 2]
-            rgb.append(int(color_hex, 16))
+        if len(color) == 3:
+            color = "".join(part * 2 for part in color)
```
Nice conversion, though a comment might help.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
+            color = "".join(part * 2 for part in color)
 
-        return rgb[0], rgb[1], rgb[2]
+        rgb = int(color, 16)
+
+        return ((rgb & 0xFF0000) >> 16, (rgb & 0x00FF00) >> 8, (rgb & 0x0000FF))
```
I appreciate that this is potentially a bit faster, but I think the readability of my solution is more useful over all. If there is a major improvement to speed I am fine with this change, however.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
     def __call__(self, text: str, color: ColorType, reset_color: bool = True) -> str:
-        """Color a piece of text using `color`.
+        """
+        Color a piece of text using `color`.
 
         The color can be one of 4 formats:
             - str colorname:   One of the predifined named colors. See the names dict.
             - int 0-256:       One of the 256 terminal colors.
             - str #RRGGBB:     CSS-style HEX string, without alpha.
             - tuple (0-256, 0-256, 0-256): Tuple of integers, representing an RGB color.
 
-        If `reset_color` is set a reset sequence is inserted at the end."""
+        If `reset_color` is set a reset sequence is inserted at the end.
+        """
```
Once again, unnecessary.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
-        # rgb values
-        if isinstance(color, tuple):
+        # 24-bit/RGB values (16581375 ≈16m colors)
+        if isinstance(color, tuple) and all([isinstance(part, int) for part in color]):
```
As pylint will tell you, there is no need to use list comprehension around a generator in `all`, as it iterates through it anyways.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
 
-        # rgb values
-        if isinstance(color, tuple):
+        # 24-bit/RGB values (16581375 ≈16m colors)
```
Not sure about the 16M comment here, but it doesn't hurt anything.

UltraStudioLTD commented:
Also I am researching ANSI mouse reports and I will update code as soon as I have working functions for both **UTF-8** (**1005 mode**) and **Urxvt (*rxvt-unicode*)** (**1015 mode**). I know formats but don't know what are event codes so I will research them. Also it would be better to change `translator` to work with regex capture groups instead of manual parsing (which is inefficient and decreases speed). But for know it will be better to merge this PR.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
-
-        # 8-bit colors (including 0-16)
-        elif isinstance(color, int) or color.isdigit():
+        # 8-bit/Xterm colors (256 colors)
+        elif int(color) in range(256):
```
There is no guarantee `int(color)` won't fail with `ValueError` here.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
             raise NotImplementedError(
                 f"Not sure what to do with {color} of type {type(color)}"
             )
+        ending = "\x1b[0m" if reset_style else ""
+        return f"\x1b[{38 + self.layer_offset};{color_value}{text}{ending}"
```
Why reimplement the `set_mode` functions here? It's more readable if we make use of the code we already have.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
 
+foreground: Color = Color(0)
+"""
+`Color` instance to setting foreground colors
+"""
 
-def screen_size() -> tuple[int, int]:
-    """Get screen size using the os module
+background: Color = Color(1)
+"""
+`Color` instance to setting background colors
+"""
```
Again, not necessary. See my comment a bit below.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
+        Initialize `TerminalModes` class
+        """
+        if _IS_NT:
+            self._OldStdinMode = ctypes.wintypes.DWORD()
+            self._OldStdoutMode = ctypes.wintypes.DWORD()
```
`_OldStdinMode` and `_OldStdoutMode` are not a valid names in the `snake_case` definition. Pylint will complain here.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
-    try:
-        width, height = get_terminal_size()
-        return (width, height)
+class TerminalModes:
```
Great functionality! I think we could add this to `Terminal`, though.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
 
-        raise
+    @contextmanager
+    def NoEcho(self) -> None:
```
I don't enjoy the PascalCase here. That is only supposed to be used by classes, which this is not. `no_echo` would work better.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
+            kernel32.SetConsoleMode(kernel32.GetStdHandle(-10), 0)
+            kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), 7)
+        else:
+            _ = self._OldStdinMode
```
As I will explain below, I don't think `_` is a good variable name.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
+            kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), 7)
+        else:
+            _ = self._OldStdinMode
+            _[3] = _[3] & ~(termios.ECHO | termios.ICANON)
+            termios.tcsetattr(sys.stdin, termios.TCSAFLUSH, _)
+        try:
+            yield
+        finally:
+            if _IS_NT:
+                kernel32.SetConsoleMode(kernel32.GetStdHandle(-10), self._OldStdinMode)
+                kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), self._OldStdoutMode)
+            else:
+                termios.tcsetattr(sys.stdin, termios.TCSAFLUSH, self._OldStdinMode)
```
Isn't this functionality implemented elsewhere already? `unset_echo` does what this is doing.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
+
+
+terminal_modes = TerminalModes()
+"""
+TerminalModes instance
+"""
```
More explanation would be great, though I would prefer this functionality to exist within `_Terminal` instead.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
+class Terminal:
+    """
+    A class to store & access data about a terminal
+    """
 
     RESIZE = 0
     margins = [0, 0, 0, 0]
 
     def __init__(self) -> None:
-        """Initialize object"""
-
+        """
+        Initialize 'Terminal' class
+        """
```
Unnecessary changes to the docstring format.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
             for callback in self._listeners[event]:
                 callback(data)
 
-    def _get_size(self) -> tuple[int, int]:
-        """Get screen size while substracting the origin position"""
-
-        # This always has len() == 2, but mypy can't see that.
-        return tuple(val - org for val, org in zip(screen_size(), self.origin))  # type: ignore
```
The `origin` calculations here are not done if `get_terminal_size` is used. You should **not** remove this.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
 
-    def _call_listener(self, event: int, data: Any) -> None:
-        """Call event callback is one is found."""
+    async def _WinSIGWINCH(self) -> None:
+        while True:
+            n_width = get_terminal_size()
+            if n_width != self.size:
+                self._update_size(28, (sys._getframe(1) if _SYS_HAS_FRAME else None))
+            await asyncio.sleep(0.5)
```
As mentioned above, this is OK, but I'm not certain on the async nature of it.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
-        # This always has len() == 2, but mypy can't see that.
-        return tuple(val - org for val, org in zip(screen_size(), self.origin))  # type: ignore
-
-    def _update_size(self, *_: Any) -> None:
-        """Resize terminal when SIGWINCH occurs.
-
-        Note:
-            SIGWINCH is not supported on Windows, so this isn't called."""
-
-        self.size = self._get_size()
+    def _update_size(self, signum: int, frame: FrameType) -> None:
+        """
+        Resize terminal when SIGWINCH or _WinSIGWINCH occurs
+        """
```
Docstring changes, again.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
-        """Subscribe a callback function to an event
+        """
+        Subscribe a callback function to an event
 
         The callback takes an event-specific argument payload:
-            RESIZE: tuple[int, int] - New screen size"""
-
+            RESIZE: tuple[int, int] - New screen size
+        """
```
I commented my piece on the docstring formats further down the page, and it's getting a bit cumbersome saying "coming soon" in every docstring of the file.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
         for height in range(self.height):
-            sys.stdout.write(
-                f"\033[{height};0H" + background(" " * (self.width - 1), color)
-            )
-
+            sys.stdout.write(f"\x1b[{height};0H{fill}")
```
Interesting change, I would use `buffer` or something along those lines instead of `fill` for the name, though.

Also not really sure if this is any better in terms of speed.

UltraStudioLTD replied:
I used `_fill` without thinking 😅. I was in hurry so I didn't checked properly and added random variable name in speed

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -107,243 +142,263 @@ class Color:
 
-class _Terminal:
-    """A class to store & access data about a terminal"""
+class Terminal:
```
There is a good reason this class was not public, and is not exported in `__all__`. It is supposed to be purely internal, with the only instance of it being `terminal`.

UltraStudioLTD replied:
Ok.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -364,146 +418,175 @@ def clear(what: str = "screen") -> None:
+    Move cursor to `pos`.
 
     Note:
         This does not flush the terminal for performance reasons. You
-        can do it manually with `sys.stdout.flush()`."""
-
-    posx, posy = pos
-    _stdout.write(f"\x1b[{posy};{posx}H")
+        can do it manually with `sys.stdout.flush()`.
+    """
```
See above for the docstring shenanigans.

bczsalba commented on pytermgui/ansi_interface.py:
```diff
@@ -364,146 +418,175 @@ def clear(what: str = "screen") -> None:
+    """
+    Move cursor up by `num` lines.
 
     Note:
         This does not flush the terminal for performance reasons. You
-        can do it manually with `sys.stdout.flush()`."""
-
-    _stdout.write(f"\x1b[{num}A")
+        can do it manually with `sys.stdout.flush()`.
+    """
+    sys.stdout.write(f"\x1b[{num}A")
```
This applies to all the changes below as well.

bczsalba commented on pytermgui/cmd.py:
```diff
@@ -10,35 +10,20 @@
-    markup,
-    Window,
-    Button,
-    Label,
-    boxes,
-    keys,
-)
+from . import Label  # prettify_markup,
+from . import (Button, Container, InputField, MarkupFormatter,
+               MarkupSyntaxError, Splitter, Window, WindowManager, YamlLoader,
+               __version__, boxes, get_widget, keys, markup, real_length,
+               terminal)
```
I don't agree with these changes, `black` will modify it as well.

I would also argue it is a bit odd to make formatting changes to all files in a library, even ones you didn't manually touch, but that's kind of on the formatter.

bczsalba commented on pytermgui/enums.py:
```diff
@@ -16,8 +16,9 @@
 
 from __future__ import annotations
 
+from enum import IntEnum
+from enum import auto as _auto
```
Not really sure if this is any better, but it is out of the scope of this PR.

UltraStudioLTD replied:
I only touched `ansi_interface` manually. All others are after `black` and `isort`'s usage

bczsalba replied:
Next time it would be better to only commit the file you are changing.

bczsalba reviewed (changes requested):
You made a couple of nice changes, but the amount of meddling the formatting has received makes this unmergeable in its current state. A lot of this isn't on you, as there is no explicit guide on how to contribute, but the issues are still there.

There is also way too many changes happening here. In my opinion, this PR should have ended with adding the `WinSIGWINCH` utility. You could've made another for the `TerminalModes` class, and another for general formatting.

My suggestion: Focus on what (I think) you intended on addressing, the Windows `SIGWINCH` call. Leave all unrelated changes for another day.

bczsalba commented:
> `Terminal` (not `_Terminal`. Please remove `_` during imports and in variables/classes which are intended to be public)

As I explain in my review, `_Terminal` is NOT intended to be public, `terminal` is. I don't really appreciate the tone of voice here either.

UltraStudioLTD commented:
> > `Terminal` (not `_Terminal`. Please remove `_` during imports and in variables/classes which are intended to be public)
> 
> As I explain in my review, `_Terminal` is NOT intended to be public, `terminal` is. I don't really appreciate the tone of voice here either.
> 
> > Also it would be better to change translator to work with regex capture groups instead of manual parsing (which is inefficient and decreases speed)
> 
> What do you mean here? It already uses regex, the only part that doesn't is the `M` suffix which is only present in one subtype.
> 
> > But for know it will be better to merge this PR.
> 
> As you can probably tell from my review above I don't really think so. I'm not really sure why you use such declarative language in these comments, saying "it _might_ be better" sounds a lot less demanding.

Sorry if I sounded declarative but I didn't meant to. My English knowledge is not that high so I use translator occasionally which doesn't always yield how I want to (but what choice do I have 😅) say. Also this part here

> But for know it will be better to merge this PR.

This is mistake of translator. I wanted to say following (I don't know if this is correct translation)

> I think this PR after merging will better it.

Yes it sounds so much different but that's how bad translator is. **Again, sorry if I sounded declarative 😅 but please know that I didn't meant to **

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -364,146 +418,175 @@ def clear(what: str = "screen") -> None:
+        kernel32.GetConsoleMode(kernel32.GetStdHandle(-11), ctypes.byref(OldStdoutMode))
+        kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), 7)
+    else:
+        _ = OldStdinMode = termios.tcgetattr(sys.stdin)
```
About that I think `tmp` will sound better

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -364,146 +418,175 @@ def clear(what: str = "screen") -> None:
-    _stdout.write("\x1b[s")
+    Use `restore_cursor()` to restore it.
+    """
+    sys.stdout.write("\x1b[s")
 
 
 def restore_cursor() -> None:
-    """Restore cursor position saved by `save_cursor()`"""
-
-    # _tput(['rc'])
-    _stdout.write("\x1b[u")
-
-
-def report_cursor() -> tuple[int, int] | None:
-    """Get position of cursor"""
-
-    print("\x1b[6n")
-    chars = getch()
-    posy, posx = chars[2:-1].split(";")
+    """
+    Restore cursor position saved by `save_cursor()`
+    """
+    sys.stdout.write("\x1b[u")
```
Alright 👍

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -364,146 +418,175 @@ def clear(what: str = "screen") -> None:
-
-    # _tput(['civis'])
-    print("\x1b[?25l")
+    """
+    Stop printing cursor
+    """
```
👍

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -364,146 +418,175 @@ def clear(what: str = "screen") -> None:
+        kernel32.GetConsoleMode(kernel32.GetStdHandle(-11), ctypes.byref(OldStdoutMode))
+        kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), 7)
+    else:
+        _ = OldStdinMode = termios.tcgetattr(sys.stdin)
+        _[3] = _[3] & ~(termios.ECHO | termios.ICANON)
+        termios.tcsetattr(sys.stdin, termios.TCSAFLUSH, _)
```
It is needed here. 😅 I will add comments later.

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -518,94 +601,94 @@ def set_mode(mode: Union[str, int], write: bool = True) -> str:
+    using one of the specific setters, like `bold` or `italic`.
+    """
+    if isinstance(mode, str) and (not mode.isdigit()):
+        mode = DECORATION_MODES.get(mode, None)
```
Ok

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -518,94 +601,94 @@ def set_mode(mode: Union[str, int], write: bool = True) -> str:
-        return
-
-    system("stty echo")
+    """
+    Start echoing user input
+    """
+    if _IS_NT:
+        kernel32.SetConsoleMode(kernel32.GetStdHandle(-10), 0)
+        kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), 7)
+    else:
+        _ = terminal_modes._OldStdinMode
+        _[3] = _[3] & ~(termios.ECHO | termios.ICANON)
+        termios.tcsetattr(sys.stdin, termios.TCSAFLUSH, _)
```
It is better to stick standard library than rely on outside tools

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -518,94 +601,94 @@ def set_mode(mode: Union[str, int], write: bool = True) -> str:
-    Note: This only gets registered when hover events are listened to"""
+    LEFT_CLICK = auto()
+    """
+    Start of a left button action sequence
+    """
+    LEFT_DRAG = auto()
+    """
+    Mouse moved while left button was held down
+    """
+    RIGHT_CLICK = auto()
+    """
+    Start of a right button action sequence
+    """
+    RIGHT_DRAG = auto()
+    """
+    Mouse moved while right button was held down
+    """
+    SCROLL_UP = auto()
+    """
+    Mouse wheel or touchpad scroll upwards
+    """
+    SCROLL_DOWN = auto()
+    """
+    Mouse wheel or touchpad scroll downwards
+    """
+    HOVER = auto()
+    """
+    Mouse moved without clicking
 
-    RELEASE = _auto()
-    """Mouse button released; end of any and all mouse action sequences"""
+    Note: This only gets registered when hover events are listened to
+    """
+    RELEASE = auto()
```
Ok

UltraStudioLTD commented on pytermgui/ansi_interface.py:
```diff
@@ -518,94 +601,94 @@ def set_mode(mode: Union[str, int], write: bool = True) -> str:
+    Its first argument is a `MouseAction` describing what happened, and its second argument is a `tuple[int, int]` describing where it happened.
 
-    This class mostly exists for readability & typing reasons. It also
-    implements the iterable protocol, so you can use the unpacking syntax,
-    such as:
+    This class mostly exists for readability & typing reasons. It also implements the iterable protocol, so you can use the unpacking syntax, such as:
```
I don't have `pylint` so I didn't checked it and I thought that you accidentally split line

bczsalba commented:
That's alright! I still stand by my statements on the code in general, however.

UltraStudioLTD commented:
Alright I will close this PR and open new but in parts.
First will be `WinSIGWINCH` implementation (but it will stay asynchronous)
Second: `TerminalModes` which will be merged with `_Terminal`.
Third: Better `Color`
Then optimisation of others

bczsalba commented:
That would be perfect! I'd prefer if you didn't run `isort` on the code, as it creates a lot of changes that I will not merge, so it's easier for everyone.

UltraStudioLTD commented:
Alright

UltraStudioLTD commented:
What about `black`? Also I will use `pylint` this time

bczsalba commented:
You should use `black`, thank you!
